### ISSUE TITLE
cross-browser differences when selecting a paragraph and trying to type

### ISSUE THREAD

**Issue by takhtamysh** [ianstormtaylor/slate#3387]
#### Do you want to request a _feature_ or report a _bug_?

A bug

#### What's the current behavior?

Currently, when selecting a paragraph using triple-click and trying to type, the text will be moved to the next paragraph. This behaviour can be reproduced on Safari and Chrome.
![chrome](https://user-images.githubusercontent.com/45436118/71446606-363e6f00-2736-11ea-9b89-760a19dc5a41.gif)

Slate: 0.57.1
Browser: Chrome / Safari
OS: Mac

#### What's the expected behavior?

The typed text should be written in the same paragraph that was selected. As works in Firefox:
![firefox](https://user-images.githubusercontent.com/45436118/71446615-43f3f480-2736-11ea-971e-e5c223555ea9.gif)

**Comment by Dundercover** [comment]
Is this really a bug? The problem revolves around if the line break is included or not in the selection which it is in chrome and safari when triple-clicking but not in firefox. Most RTEs have this behaviour I think. The comment field here on github which I'm writing this comment in shows the same behaviour.

**Comment by eshavrov** [comment]
I agree with your statement, but the selected fragment should be deleted

**Comment by Dundercover** [comment]
Ah, yeah, you don't want that lingering empty block above :)

**Comment by anges244** [comment]
This is a more general problem with triple-click selection on Chrome. You can replicate it while trying to select the whole first paragraph in the images example by clicking 3 times in a word. In Safari, it seems to work as expected.

![selection_bug_slate](https://user-images.githubusercontent.com/3290236/92045701-dd75aa80-ed89-11ea-9295-4ffe3e1df580.gif)

Any ideas on how to handle that? Tried with the onDoubleClick but had no success.

**Comment by eshavrov** [comment]
This is a hack, not a solution))
Since it was critical for me. I had to fix it like this.
```javascript
const selectOneAllElement = ({ anchor, focus }) => {
  if (anchor.offset !== 0 || focus.offset !== 0) {
    return false;
  }

  const anchorPath = anchor.path.slice();
  const focusPath = focus.path.slice();

  if (anchorPath.length === 0 || focusPath.length === 0) {
    return false;
  }

  while (anchorPath.length > 0 && anchorPath[0] === focusPath[0]) {
    anchorPath.shift();
    focusPath.shift();
  }

  if (anchorPath.length > 0 || focusPath.length > 0) {
    return anchorPath[0] + 1 === focusPath[0];
  }

  return false;
};

const fixAt = editor => {
  const { selection } = editor;

  if (Range.isRange(selection)) {
    const { anchor } = selection;

    if (selectOneAllElement(selection)) {
      const end = Editor.end(editor, anchor);

      return { anchor, focus: end };
    }
  }

  return selection;
};

const hackSelection = editor => {
  const fixSelection = fixAt(editor);

  if (fixSelection) {
    Transforms.select(editor, fixSelection);
  }
};
```
This is a temporary solution that is planned to be removed when it is fixed in the slate

**Comment by anges244** [comment]
@eshavrov thank you! Definitely a hack but much appreciated. Combined it with an isChrome detection and it should work until fixed by default.

**Comment by anges244** [comment]
Building on the solution of @eshavrov, this seems to be a sufficient but hacky solution for anybody who has the same problem.

```
import {Editor, Transforms, Range, Element } from 'slate';
import { ReactEditor } from 'slate-react';
const isChrome = !!window.chrome && (!!window.chrome.webstore || !!window.chrome.runtime);

export const checkAndFixChromeSelection = (editor) => {
	if (!isChrome){ return; } // Apply only on Chrome
	const { selection } = editor;
	if (Range.isRange(selection) && !Range.isCollapsed(selection)){
		// Detect if next block is selected
		const { anchor, focus } = selection;
		if (anchor.offset === 0 && focus.offset === 1 && anchor.path[0] !== focus.path[0]){
			// Find the selected nodes
			const [nodes] = Editor.nodes(editor, { at: selection, 	match: n => Element.isElement(n)});
			// Find the node path
			const path = ReactEditor.findPath(editor,nodes[0]);
			// End of the current node
			const end = Editor.end(editor, path);
			// Select whole node
			return Transforms.select(editor, { anchor, focus: end });
		}
	}
} 
```
Running this function on the Slate component seems to work fine.

**Comment by jeremysimms** [comment]
> Building on the solution of @eshavrov, this seems to be a sufficient but hacky solution for anybody who has the same problem.
> 
> ```
> import {Editor, Transforms, Range, Element } from 'slate';
> import { ReactEditor } from 'slate-react';
> const isChrome = !!window.chrome && (!!window.chrome.webstore || !!window.chrome.runtime);
> 
> export const checkAndFixChromeSelection = (editor) => {
> 	if (!isChrome){ return; } // Apply only on Chrome
> 	const { selection } = editor;
> 	if (Range.isRange(selection) && !Range.isCollapsed(selection)){
> 		// Detect if next block is selected
> 		const { anchor, focus } = selection;
> 		if (anchor.offset === 0 && focus.offset === 1 && anchor.path[0] !== focus.path[0]){
> 			// Find the selected nodes
> 			const [nodes] = Editor.nodes(editor, { at: selection, 	match: n => Element.isElement(n)});
> 			// Find the node path
> 			const path = ReactEditor.findPath(editor,nodes[0]);
> 			// End of the current node
> 			const end = Editor.end(editor, path);
> 			// Select whole node
> 			return Transforms.select(editor, { anchor, focus: end });
> 		}
> 	}
> } 
> ```
> 
> Running this function on the Slate component seems to work fine.

Thanks for the code snippet. Are you attaching this to an event handler or using it as a plugin?
